File: segundo/1er cuatri/PDOO/Irrgarten_java/src/irrgarten/Labyrinth.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package irrgarten;
import java.util.ArrayList;
/**
 *
 * @author alvarogz
 */
public class Labyrinth {
    static private final char BLOCK_CHAR = 'X';
    static private final char EMPTY_CHAR = '-';
    static private final char MONSTER_CHAR = 'M';
    static private final char COMBAT_CHAR = 'C';
    static private final char EXIT_CHAR = 'E';
    static private final int ROW = 0;
    static private final int COL = 1;
    
    private int nRows;
    private int nCols;
    private int exitRow;
    private int exitCol;
    
    private Monster[][] monsters;
    private Player[][] players;
    private char[][] laberinto;
     
    public Labyrinth(int nRows, int nCols, int exitRow, int exitCol) {
        this.exitRow = exitRow;
        this.exitCol = exitCol;
        this.nRows = nRows;
        this.nCols = nCols;
        
        monsters = new Monster[this.nRows][this.nCols];
        players = new Player[this.nRows][this.nCols];
        laberinto = new char[this.nRows][this.nCols];
        
        for (int i = 0; i < this.nRows; i++){
            for (int j = 0; j < this.nCols; j++){
                laberinto[i][j] = EMPTY_CHAR;
            }
        }
        
         laberinto[exitRow][exitCol] = EXIT_CHAR;
    }
    
    public void spreadPlayers(ArrayList<Player> players, boolean modoDebug) {
        if (!modoDebug){   
            for (Player p : players){
                int pos[] = randomEmptyPos();

                int oldRow = -1;
                int oldCol = -1;
                putPlayer2D(oldRow, oldCol, pos[ROW], pos[COL], p);
            }
        }
        else {
            Player p = players.get(0);
            this.putPlayer2D(-1, -1, 0, 0, p);
        }
    }
    
    public boolean haveAWinner() {
        return laberinto[this.exitRow][this.exitCol] != EXIT_CHAR;
    }
    
    @Override
    public String toString() {
        String laux = "Tablero: \n";
        
        for (int i = 0; i < this.nRows; i++){
            for (int j = 0; j < this.nCols; j++){
                laux += laberinto[i][j] + "   ";
            }
            
            laux += "\n";
        }
        
        return laux;
    }
    
    public void addMonster(int row, int col, Monster monster) {
        if (posOk(row, col) && emptyPos(row, col)){
            laberinto[row][col] = MONSTER_CHAR;
            this.monsters[row][col] = monster;
        }
    }
    
    public Monster putPlayer(Directions direction, Player player) {
        int oldRow = player.getRow();
        int oldCol = player.getCol();
        
        int newPos[] = dir2Pos(oldRow, oldCol, direction);
        Monster monster = putPlayer2D(oldRow, oldCol, newPos[ROW], newPos[COL], player);
        
        return monster;
    }
    
    public void addBlock(Orientation orientation, int startRow, int startCol, int length) {
        int incRow;
        int incCol;
        int row;
        int col;
        
        if (orientation == Orientation.VERTICAL){
            incRow = 1;
            incCol = 0;
        }
        else {
            incRow = 0;
            incCol = 1;
        }
        
        row = startRow;
        col = startCol;
        
        while (posOk(row, col) && emptyPos(row, col) && length > 0){
            laberinto[row][col] = BLOCK_CHAR;
            length -= 1;
            row += incRow;
            col += incCol;
        }
    }
    
    public ArrayList<Directions> validMoves(int row, int col) {
        ArrayList<Directions> output = new ArrayList<>();
        
        if (canStepOn(row+1, col))
            output.add(Directions.DOWN);
        if (canStepOn(row-1, col))
            output.add(Directions.UP);
        if (canStepOn(row, col+1))
            output.add(Directions.RIGHT);
        if (canStepOn(row, col-1))
            output.add(Directions.LEFT);
                    
        return output;
    }
    
    public void colocaFuzzyPlayer(FuzzyPlayer fuzzy, int row, int col){
        players[row][col] = fuzzy;
    }
        
    private boolean posOk(int row, int col) {
        return 0 <= row && row < this.nRows && 0 <= col && col < this.nCols;
    } 
    
    private boolean emptyPos(int row, int col) {
        return laberinto[row][col] == EMPTY_CHAR;
    }
    
    private boolean monsterPos(int row, int col) {
        return laberinto[row][col] == MONSTER_CHAR;
    }
    
    private boolean exitPos(int row, int col) {
        return laberinto[row][col] == EXIT_CHAR;
    }
    
    private boolean combatPos(int row, int col) {
        return laberinto[row][col] == COMBAT_CHAR;
    }
    
    private boolean canStepOn(int row, int col) {
        return posOk(row, col) && (emptyPos(row, col) || monsterPos(row, col) || exitPos(row, col));
    }
    
    private void updateOldPos(int row, int col) {
        if (posOk(row, col)){
            if (laberinto[row][col] == COMBAT_CHAR)
                laberinto[row][col] = MONSTER_CHAR;
            else
                laberinto[row][col] = EMPTY_CHAR;
        }
    }
    
    private int[] dir2Pos(int row, int col, Directions direction) {
        int[] pos = {row, col}; //creamos una variable auxiliar con nuestra posició
        
        switch (direction) {
            case UP:
                pos[ROW]--;
                break;
            case DOWN:
                pos[ROW]++;
                break;
            case RIGHT:
                pos[COL]++;
                break;
            case LEFT:
                pos[COL]--;
                break;
        }
        
        return pos;
    }
    
    public int[] randomEmptyPos() {
        int[] pos = {-1, -1};
        
        do {
            pos[ROW] = Dice.randomPos(nRows);
            pos[COL] = Dice.randomPos(nCols);
        }while (!emptyPos(pos[ROW], pos[COL]));
        
        return pos;
    }
    
    private Monster putPlayer2D(int oldRow, int oldCol, int row, int col, Player player) {
        Monster output = null;
        
        if (canStepOn(row, col)){
            if (posOk(oldRow, oldCol)){
                Player p = players[oldRow][oldCol];
                
                if (p == player){
                    updateOldPos(oldRow, oldCol);
                    players[oldRow][oldCol] = null;
                }
            }
            
            if (monsterPos(row, col)){
                laberinto[row][col] = COMBAT_CHAR;
                output = monsters[row][col];
            }
            else {
                char number = player.getNumber();
                laberinto[row][col] = number;               
            }
            
            players[row][col] = player;
            player.setPos(row, col);
        }
        
        return output;
    }
}

File: segundo/1er cuatri/PDOO/irrgarten_ruby/Irrgarten/labyrinth.rb
# frozen_string_literal: true
require_relative 'orientation'
require_relative 'directions'
require_relative 'dice'

class Labyrinth
  @@BLOCK_CHAR = 'X'
  @@EMPTY_CHAR = '-'
  @@MONSTER_CHAR = 'M'
  @@COMBAT_CHAR = 'C'
  @@EXIT_CHAR = 'E'
  @@ROW = 0
  @@COL = 1

  def initialize(n_rows, n_cols, exit_row, exit_col)
    @n_rows = n_rows
    @n_cols = n_cols
    @exit_row = exit_row
    @exit_col = exit_col

    @monsters = Array.new(@n_rows) { Array.new(@n_cols) } #Array.new(n_cols, 0) daria un valor 0 a tpdas las casillas
    @players = Array.new(@n_rows) { Array.new(@n_cols) }
    @laberinto = Array.new(@n_rows) { Array.new(@n_cols, @@EMPTY_CHAR) }
    @laberinto[@exit_row][@exit_col] = @@EXIT_CHAR
  end

  def spread_players(players, modo_debug)
    unless modo_debug
      for p in players
        pos = random_empty_pos

        old_row = -1
        old_col = -1
        put_player2d(old_row, old_col, pos[@@ROW], pos[@@COL], p)
      end
    end
    if modo_debug
      p = players[0]
      put_player2d(-1, -1, 0, 0, p)
    end
  end

  def have_a_winner
    true if @laberinto[@exit_row][@exit_col] != @@EXIT_CHAR
  end

  def to_s
    laux = "Tablero: \n"

    for i in 0..@n_rows-1 #si solo pones .. incluye el límite
      for j in 0..@n_cols-1
        laux += "#{@laberinto[i][j]}   "
      end

      laux += "\n"
    end

    laux
  end

  def add_monster(row, col, monster)
    if pos_ok(row, col) && empty_pos(row, col)
      @laberinto[row][col] = @@MONSTER_CHAR
      @monsters[row][col] = monster
    end
  end

  def put_player(direction, player)
    old_row = player.row
    old_col = player.col

    new_pos = dir2pos(old_row, old_col, direction)
    monster = put_player2d(old_row, old_col, new_pos[@@ROW], new_pos[@@COL], player)
    monster
  end

  def coloca_fuzzy_player(fuzzy, row, col)
    @players[row][col] = fuzzy
  end

  def add_block(orientation, start_row, start_col, length)
    if orientation == Orientation::VERTICAL
      inc_row = 1
      inc_col = 0
    else
      inc_row = 0
      inc_col = 1
    end

    row = start_row
    col = start_col

    while pos_ok(row, col) && empty_pos(row, col) && length > 0
      @laberinto[row][col] = @@BLOCK_CHAR
      length -= 1
      row += inc_row
      col += inc_col
    end
  end

  def valid_moves(row, col)
    output = Array.new

    if can_step_on(row+1, col)
      output.push(Directions::DOWN)
    end
    if can_step_on(row-1, col)
      output.push(Directions::UP)
    end
    if can_step_on(row, col+1)
      output.push(Directions::RIGHT)
    end
    if can_step_on(row, col-1)
      output.push(Directions::LEFT)
    end

    output
  end

  def pos_ok(row, col)
    true if 0 <= row && row < @n_rows && 0 <= col && col < @n_cols
  end

  def empty_pos(row, col)
    true if @laberinto[row][col] == @@EMPTY_CHAR
  end

  def monster_pos(row, col)
    true if @laberinto[row][col] == @@MONSTER_CHAR
  end

  def exit_pos(row, col)
    true if @laberinto[row][col] == @@EXIT_CHAR
  end

  def combat_pos(row, col)
    true if @laberinto[row][col] == @@COMBAT_CHAR
  end

  def can_step_on(row, col)
    true if pos_ok(row, col) && (empty_pos(row, col) || monster_pos(row, col) || exit_pos(row, col))
  end

  def update_old_pos(row, col)
    if pos_ok(row, col)
      if @laberinto[row][col] == @@COMBAT_CHAR
        @laberinto[row][col] = @@MONSTER_CHAR
      else
        @laberinto[row][col] = @@EMPTY_CHAR
      end
    end
  end

  def dir2pos(row, col, direction)
    pos = [row, col]

    case direction
    when Directions::UP
      pos[@@ROW] -= 1
    when Directions::DOWN
      pos[@@ROW] += 1
    when Directions::RIGHT
      pos[@@COL] += 1
    when Directions::LEFT
      pos[@@COL] -= 1
    end

    pos
  end

  def random_empty_pos
    pos = [-1, -1]

    begin
      pos[@@ROW] = Dice.random_pos(@n_rows)
      pos[@@COL] = Dice.random_pos(@n_cols)
    end while !empty_pos(pos[@@ROW], pos[@@COL])

    pos
  end

  def put_player2d(old_row, old_col, row, col, player)
    output = nil

    if can_step_on(row, col)
      if pos_ok(old_row, old_col)
        p = @players[old_row][old_col]

        if p == player
          update_old_pos(old_row, old_col)
          @players[old_row][old_col] = nil
        end
      end

      if monster_pos(row, col)
        @laberinto[row][col] = @@COMBAT_CHAR
        output = @monsters[row][col]
      else
        number = player.number
        @laberinto[row][col] = number
      end

      @players[row][col] = player
      player.set_pos(row, col)
    end

    output
  end

  private :pos_ok, :empty_pos, :monster_pos, :exit_pos, :combat_pos, :can_step_on, :update_old_pos, :dir2pos, :put_player2d

end
